heron/metricscachemgr/src/java/com/twitter/heron/metricscachemgr/metricscache/MetricsCacheQueryUtils.java

```diff
@@ -0,0 +1,293 @@
+
+      long interval = request.getInterval(); // in seconds
+      if (interval <= 0) { // means all
+        outRequest.startTime = 0;
+      } else { // means [-interval, now]
+        outRequest.startTime = outRequest.endTime - interval * 1000;
+      }
+    } else {
+      outRequest.startTime = request.getExplicitInterval().getStart() * 1000;
+      outRequest.endTime = request.getExplicitInterval().getEnd() * 1000;
+    }
+
+    // default: aggregate all metrics
+    outRequest.aggregationGranularity = 0;
+    if (request.hasMinutely()) {
+      outRequest.aggregationGranularity = request.getMinutely() ? 1 : 0;
+    }
+
+    return outRequest;
+  }
+
+  /**
+   * compatible with com.twitter.heron.proto.tmaster.TopologyMaster.MetricResponse
+   */
+  public static TopologyMaster.MetricResponse toProtobuf(MetricResponse response,
+                                                         MetricRequest request) {
+    TopologyMaster.MetricResponse.Builder builder =
+        TopologyMaster.MetricResponse.newBuilder();
+    builder.setInterval((request.endTime - request.startTime) / 1000); // in seconds
+
+    // default OK if we have response to build already
+    builder.setStatus(Common.Status.newBuilder().setStatus(Common.StatusCode.OK));
+
+    // instanceId -> [metricName -> metricValue]
+    // componentName is ignored, since there is only one component in the query
+    Map<String, Map<String, List<MetricTimeRangeValue>>> aggregation =
+        new HashMap<>();
+    for (MetricDatum datum : response.metricList) {
+      String instanceId = datum.instanceId;
+      String metricName = datum.metricName;
+      List<MetricTimeRangeValue> metricValue = datum.metricValue;
+      // prepare
+      if (!aggregation.containsKey(instanceId)) {
+        aggregation.put(instanceId, new HashMap<String, List<MetricTimeRangeValue>>());
+      }
+      if (!aggregation.get(instanceId).containsKey(metricName)) {
+        aggregation.get(instanceId).put(metricName, new ArrayList<MetricTimeRangeValue>());
+      }
+      // aggregate
+      aggregation.get(instanceId).get(metricName).addAll(metricValue);
+    }
+
+    // add TaskMetric
+    for (String instanceId : aggregation.keySet()) {
+      TopologyMaster.MetricResponse.TaskMetric.Builder taskMetricBuilder =
+          TopologyMaster.MetricResponse.TaskMetric.newBuilder();
+
+      taskMetricBuilder.setInstanceId(instanceId);
+      // add IndividualMetric
+      for (String metricName : aggregation.get(instanceId).keySet()) {
+        TopologyMaster.MetricResponse.IndividualMetric.Builder individualMetricBuilder =
+            TopologyMaster.MetricResponse.IndividualMetric.newBuilder();
+
+        individualMetricBuilder.setName(metricName);
+        // add value|IntervalValue
+        List<MetricTimeRangeValue> list = aggregation.get(instanceId).get(metricName);
+        if (list.size() == 1) {
+          LOG.info("get0 " + list.get(0) + "; value " + list.get(0).value);
```

Comment:
This log message without add'l context will not be helpful. please remove or add context.